system/lib/wasmfs/file.cpp

```diff
@@ -40,33 +40,180 @@ void DataFile::Handle::preloadFromJS(int index) {
 // Directory
 //
 
+void Directory::Handle::finishInsertingChild(const std::string& name,
+                                             std::shared_ptr<File> child) {
+  // Update the dcache and set the child's parent.
+  auto& dcache = getDir()->dcache;
+  auto [_, inserted] = dcache.insert({name, {DCacheKind::Normal, child}});
+  assert(inserted && "inserted child already existed!");
+  assert(child->locked().getParent() == nullptr);
+  child->locked().setParent(getDir());
+}
+
+std::shared_ptr<File> Directory::Handle::getChild(const std::string& name) {
+  // Unlinked directories must be empty, without even "." or ".."
+  if (!getParent()) {
+    return nullptr;
+  }
+  if (name == ".") {
+    return file;
+  }
+  if (name == "..") {
+    return getParent();
+  }
+  // Check whether the cache already contains this child.
+  auto& dcache = getDir()->dcache;
+  if (auto it = dcache.find(name); it != dcache.end()) {
+    return it->second.file;
+  }
+  // Otherwise check whether the backend knows about the child.
+  auto child = getDir()->getChild(name);
+  if (!child) {
+    return nullptr;
+  }
+  finishInsertingChild(name, child);
+  return child;
+}
+
+bool Directory::Handle::mountChild(const std::string& name,
+                                   std::shared_ptr<File> child) {
+  assert(child);
+  // Cannot insert into an unlinked directory.
+  if (!getParent()) {
```

Comment:
Doesn't the root have no parent?

Reply:
No, I "fixed" that at some point and now the root is its own parent. That means we don't need to special case the root in situations like this.

Reply:
That change was part of #16409.